// weixin-apis/template.js
'use strict';

function now() {
    return Math.round(Date.now() / 1000);
}

module.exports = {

    text: function(msg) {
        return '<xml>' +
            '<ToUserName><![CDATA[' + msg.toUserName + ']]></ToUserName>' +
            '<FromUserName><![CDATA[' + msg.fromUserName + ']]></FromUserName>' +
            '<CreateTime>' + now() + '</CreateTime>' +
            '<MsgType><![CDATA[text]]></MsgType>' +
            '<Content><![CDATA[' + msg.content + ']]></Content>' +
            '</xml>';
    },

    image: function(msg) {
        return '<xml>' +
            '<ToUserName><![CDATA[' + msg.toUserName + ']]></ToUserName>' +
            '<FromUserName><![CDATA[' + msg.fromUserName + ']]></FromUserName>' +
            '<CreateTime>' + now() + '</CreateTime>' +
            '<MsgType><![CDATA[image]]></MsgType>' +
            '<Image>' +
            '<MediaId><![CDATA[' + msg.mediaId + ']]></MediaId>' +
            '</Image>' +
            '</xml>';
    },

    voice: function(msg) {
        return '<xml>' +
            '<ToUserName><![CDATA[' + msg.toUserName + ']]></ToUserName>' +
            '<FromUserName><![CDATA[' + msg.fromUserName + ']]></FromUserName>' +
            '<CreateTime>' + now() + '</CreateTime>' +
            '<MsgType><![CDATA[voice]]></MsgType>' +
            '<Voice>' +
            '<MediaId><![CDATA[' + msg.mediaId + ']]></MediaId>' +
            '</Voice>' +
            '</xml>';
    },

    video: function(msg) {
        return '<xml>' +
            '<ToUserName><![CDATA[' + msg.toUserName + ']]></ToUserName>' +
            '<FromUserName><![CDATA[' + msg.fromUserName + ']]></FromUserName>' +
            '<CreateTime>' + now() + '</CreateTime>' +
            '<MsgType><![CDATA[video]]></MsgType>' +
            '<Video>' +
            '<MediaId><![CDATA[' + msg.mediaId + ']]></MediaId>' +
            '<Title><![CDATA[' + msg.title + ']]></Title>' +
            '<Description><![CDATA[' + msg.description + ']]></Description>' +
            '</Video>' +
            '</xml>';
    },

    music: function(msg) {
        return '<xml>' +
            '<ToUserName><![CDATA[' + msg.toUserName + ']]></ToUserName>' +
            '<FromUserName><![CDATA[' + msg.fromUserName + ']]></FromUserName>' +
            '<CreateTime>' + now() + '</CreateTime>' +
            '<MsgType><![CDATA[music]]></MsgType>' +
            '<Music>' +
            '<Title><![CDATA[' + msg.title + ']]></Title>' +
            '<Description><![CDATA[' + msg.description + ']]></Description>' +
            '<MusicUrl><![CDATA[' + msg.musicUrl + ']]></MusicUrl>' +
            '<HQMusicUrl><![CDATA[' + msg.HQMusicUrl + ']]></HQMusicUrl>' +
            '<ThumbMediaId><![CDATA[' + msg.thumbMediaId + ']]></ThumbMediaId>' +
            '</Music>' +
            '</xml>';
    }









};
